apps/sync/js/sync-engine/syncengine.js

```diff
@@ -0,0 +1,358 @@
+          throw new SyncEngine.UnrecoverableError();
+        } else {
+          throw err;
+        }
+      });
+    },
+
+    _ensureReady: function() {
+      if (this._ready) {
+        return Promise.resolve();
+      }
+      return this._syncCollection('meta').then(() => {
+        return this._getItem('meta', 'global');
+      }).then(metaGlobal => {
+        if (!this._storageVersionOK(metaGlobal)) {
+          return Promise.reject(new SyncEngine.UnrecoverableError(`Incompatible\
+ storage version or storage version not recognized.`));
+        }
+        return this._syncCollection('crypto');
+      }).then(() => {
+        return this._getItem('crypto', 'keys');
+      }).then((cryptoKeysRecord) => {
+        var cryptoKeys;
+        try {
+          cryptoKeys = JSON.parse(cryptoKeysRecord.data.payload);
+        } catch (e) {
+          return Promise.reject(new SyncEngine.UnrecoverableError(`Could not pa\
+rse crypto/keys payload as JSON`));
+        }
+        return cryptoKeys;
+      }).then((cryptoKeys) => {
+        return this._initFxSyncWebCrypto(cryptoKeys);
+      });
+    },
+
+    _createCollections: function() {
+      for (var collectionName in this._adapters) {
+        this._collections[collectionName] = this._kinto.collection(
+            collectionName);
+        this._collections[collectionName].use(new FxSyncIdSchema(
+            collectionName));
+        this._collections[collectionName].use(new WebCryptoTransformer(
+            collectionName, this._fswc));
+      }
+    },
+
+    _updateCollection: function(collectionName) {
+      return this._syncCollection(collectionName).then(() => {
+        return this._adapters[collectionName].update(
+            this._collections[collectionName]);
+      }).then(() => {
+        return this._syncCollection(collectionName);
+      });
+    },
+
+    /**
+      * syncNow - Syncs collections up and down between device and server.
+      * @param {Array of Strings} collectionNames The names of the collections
+      *                                           to sync.
+      * @returns {Promise}
+      */
+    syncNow: function(collectionNames) {
+      if (!Array.isArray(collectionNames)) {
+        return Promise.reject(new Error('collectionNames should be an Array'));
+      }
+      return this._ensureReady().then(() => {
+        var promises = [];
+        collectionNames.forEach(collectionName => {
+          promises.push(this._updateCollection(collectionName));
+        });
+        return Promise.all(promises);
+      });
+    }
+  };
+
+  SyncEngine.UnrecoverableError = function() {
+    this.message = 'unrecoverable';
+  };
+  SyncEngine.UnrecoverableError.prototype = Object.create(Error.prototype);
+
+  SyncEngine.TryLaterError = function() {
+    this.message = 'try later';
+  };
+  SyncEngine.TryLaterError.prototype = Object.create(Error.prototype);
+
+  SyncEngine.AuthError = function() {
+    this.message = 'unauthorized';
+  };
```

Comment:
We'll need to review this later when we put all the pieces together. But that can be done in a follow up.